hack/generator/pkg/astmodel/storage_conversion_factories.go

```diff
@@ -727,6 +702,374 @@ func assignOptionalEnumTypeFromOptionalEnumType(
+//
+// For ConvertFrom:
+//
+// var <local> <destinationType>
+// if <source> != nil {
+//     err := <local>.ConvertFrom(<source>)
+//     if err != nil {
+//         return errors.Wrap(err, "while calling <local>.ConvertTo(<source>)")
+//     }
+// }
+// <destination> = <local>
+//
+// For ConvertTo:
+//
+// var <local> <destinationType>
+// if <source> != nil {
+//     err := <source>.ConvertTo(&<local>)
+//     if err != nil {
+//         return errors.Wrap(err, "while calling <local>.ConvertTo(<source>)")
+//     }
+// }
+// <destination> = <local>
+//
+func assignObjectTypeFromOptionalObjectType(
+	sourceEndpoint *StorageConversionEndpoint,
+	destinationEndpoint *StorageConversionEndpoint,
+	conversionContext *StorageConversionContext) StorageTypeConversion {
+
+	// Require source to be optional
+	if _, srcOpt := AsOptionalType(sourceEndpoint.Type()); !srcOpt {
+		return nil
+	}
+
+	// Require destination to be non-optional
+	if _, dstOpt := AsOptionalType(destinationEndpoint.Type()); dstOpt {
+		return nil
+	}
+
+	// Require source to be an object
+	if _, srcIsObject := conversionContext.ResolveObject(sourceEndpoint.Type()); !srcIsObject {
+		return nil
+	}
+
+	// Require destination to be an object
+	dstName, dstIsObject := conversionContext.ResolveObject(destinationEndpoint.Type())
+	if !dstIsObject {
+		return nil
+	}
+
+	copyvar := destinationEndpoint.CreateLocal()
+
+	return func(reader dst.Expr, writer dst.Expr, generationContext *CodeGenerationContext) []dst.Stmt {
+
+		localId := dst.NewIdent(copyvar)
+		errLocal := dst.NewIdent("err")
+
+		declaration := astbuilder.LocalVariableDeclaration(copyvar, createTypeDeclaration(dstName, generationContext), "")
+
+		var conversion dst.Stmt
+		if dstName.PackageReference.Equals(generationContext.CurrentPackage()) {
+			conversion = astbuilder.SimpleAssignment(
+				errLocal,
+				token.DEFINE,
+				astbuilder.CallExpr(localId, conversionContext.functionName, reader))
+		} else {
+			conversion = astbuilder.SimpleAssignment(
+				errLocal,
+				token.DEFINE,
+				astbuilder.CallExpr(reader, conversionContext.functionName, localId))
+		}
+
+		checkForError := astbuilder.ReturnIfNotNil(
+			errLocal,
+			astbuilder.WrappedErrorf(
+				"populating %s from %s, calling %s()",
+				destinationEndpoint.name, sourceEndpoint.name, conversionContext.functionName))
+
+		safeConversion := astbuilder.IfNotNil(reader, conversion, checkForError)
+
+		assignment := astbuilder.SimpleAssignment(
+			writer,
+			token.ASSIGN,
+			dst.NewIdent(copyvar))
+
+		return []dst.Stmt{
+			declaration,
+			safeConversion,
+			assignment,
+		}
+	}
+}
+
+// assignOptionalObjectTypeFromObjectType will generate a conversion if both properties are
+// TypeNames referencing ObjectType definitions and only the destination is optional
+//
+// For ConvertFrom:
+//
+// var <local> <destinationType>
+// <local>.ConvertFrom(<source>)
+// <destination> = &<local>
```

Comment:
These are missing error handling (although as far as I can see the generated code is doing it).

Reply:
(Confirmed after looking at the golden files.)

Reply:
Might be ok to omit the error handling here for clarity of comment, but possibly worth noting that it's been done?

```
For ConvertFrom (error handling omitted for brevity):
...
```

?